Terminal.Gui/Input/Command.cs

```diff
@@ -4,10 +4,35 @@
 
 namespace Terminal.Gui {
 
+
+	/// <summary>
+	/// Defines the scope of a <see cref="Command"/> that has been bound to a key with <see cref="View.AddKeyBinding"/>.
+	/// </summary>
+	public enum CommandScope {
+		/// <summary>
+		/// The command is scoped to just the view that adds the key binding.
+		/// </summary>
+		View = 0,
+
+		/// <summary>
+		/// The command is scoped to the view that adds the key binding and that view's <see cref="View.SuperView"/>.
```

Comment:
I think that you are calling this `SuperView` because the keybinding will take effect even when the view is not in focus, so long as its parent still contains the focus? Is that correct?

If so then maybe the comment should be:

```csharp
/// <summary>
/// The keybinding will be triggered even when the View is not in focus.  So long as it's SuperView still contains focus.
/// <remarks>Use for Views such as MenuBar and StatusBar which provide commands that trigger even when not focused.</remarks>
/// </summary>
```

Reply:
@tig is doing a great work and he understand what I said about key bindings must have something to distinguish what command is sent for the view that has focused or not. In my opinion the focused view has always priority to handle any of the key bindings scope.
So I propose the following:
1. Firstly simplifying the return value of the `OnInvokingKeyBindings` to only `true` or `false` and having a method the returns the `Key` for a command scope to allow a user to check if a view contains it.
2. Adding a scope parameter to the `OnInvokingKeyBindings` for the view handle the passed scope. It can be set with a default value of 0.
3. Maybe renaming the `enum` fields scope to identify the scope. `Focus` meaning that only the most focused view can handle the command that are set with the this scope. `HotKey` meaning that the most focused view has priority to handle this scope first and if return `false` then check upwards if any view want it, even without focus. `Global` meaning that the most focused view has priority to handle this scope and if return `false` then check upwards if any view want it, even without focus.

Reply:
> @tig is doing a great work and he understand what I said about key bindings must have something to distinguish what command is sent for the view that has focused or not. In my opinion the focused view has always priority to handle any of the key bindings scope. 

Doesn't the fact that `ProcessKeyDown` first try `if (Focused?.ProcessKeyDown (keyEvent) == true)` ensure the focused view has priority? 

> So I propose the following:
> 
> 1. Firstly simplifying the return value of the `OnInvokingKeyBindings` to only `true` or `false` and having a method the returns the `Key` for a command scope to allow a user to check if a view contains it.

I'm not sure about the need for this given my assertion above.

> 2. Adding a scope parameter to the `OnInvokingKeyBindings` for the view handle the passed scope. It can be set with a default value of 0.

Regardless of my assertion above, I like this idea. However, doing so requires us to expand `KeyEventArgs` to have a `KeyBindingScope` property as well...

> 3. Maybe renaming the `enum` fields scope to identify the scope. `Focus` meaning that only the most focused view can handle the command that are set with the this scope. `HotKey` meaning that the most focused view has priority to handle this scope first and if return `false` then check upwards if any view want it, even without focus. `Global` meaning that the most focused view has priority to handle this scope and if return `false` then check upwards if any view want it, even without focus.

I'll think about this. It seems to make sense!

> 4. Thus it isn't not need to check if it's a `MenuBar` or a `StatusBar` because they will override the `OnInvokingKeyBindings` and handle them selfs the right commands.

I love this!

Reply:
> Doesn't the fact that ProcessKeyDown first try if (Focused?.ProcessKeyDown (keyEvent) == true) ensure the focused view has priority?

Yes it do, but only to run first the overridden virtual method of the derived class which also call the `OnInvokingKeyBindings` and not for the event that the most focused view subscribed. I propose this and the same for down and up with the right scopes.

```cs
	public bool ProcessKeyPressed (KeyEventArgs keyEvent)
	{
		if (!Enabled) {
			return false;
		}

		if (MostFocused != null && MostFocused == Focused) {
			// fire event
			Focused.KeyPressed?.Invoke (this, keyEvent);
			if (keyEvent.Handled) {
				return true;
			}
		}

		if (Focused?.ProcessKeyPressed (keyEvent) == true) {
			return true;
		} else if (Focused == null && ProcessKeyPressed (keyEvent)) {
			return true;
		}

		var handled = OnInvokingKeyBindings (keyEvent);
		if (handled != null && (bool)handled) {
			return true;
		}

		return false;
	}

```

```cs
	public virtual bool OnKeyPressed (KeyEventArgs keyEvent)
	{
		return false;
	}

```

> I'm not sure about the need for this given my assertion above.

Because the code is more simple and more easy. Change to non nullable `bool`.

```cs
		if (OnInvokingKeyBindings (keyEvent)) {
			return true;
		}

```

Reply:
If `Focused == null` then call the `OnKeyPressed (keyEvent)` instead of `ProcessKeyPressed (keyEvent)`.

Reply:
If `Focused == null` then call the `KeyPressed?.Invoke (this, keyEvent);` before advance forward.